### PR TITLE
Crowdsourcing data compiler class for using Mephisto abstractions.

### PR DESCRIPTION
**Pull request by mojtaba-komeili**
**Patch description**
* Cleaned up `analysis.py` module in ParlAI crowdsourcing: moved logic from `AbstractDataBrowserResultsCompiler` to `AbstractResultsCompiler`.
* Added a readme entry that explains how to extend `AbstractResultsCompiler` for compiling a dataset from a Mephisto task.
* Added new unit tests with mock versions of `MephistoDB` and `DataBrowser`.

**Testing steps**
* The newly added unit tests.
* Compiling a subset of Wizard of Internet data task with a new class extended from the `AbstractResultsCompiler` (see PS 2).

PS 1. kept `AbstractTurnAnnotationResultsCompiler` for now for not breaking any projects.
PS 2. I made a cleaned up version of wizard of internet data compiler that is based on this new `AbstractResultsCompiler`. The next step after landing this would be to add that.

### CODE REVIEW

File: parlai/crowdsourcing/utils/README.md [hunk 1]
```diff
@@ -3,7 +3,7 @@
 ## Overview
 - `acceptability.py`: Used to ensure that a worker's messages throughout a conversation meet certain criteria (not too short, not all caps, not a lot of repetition, safety, etc.). More details about the acceptability checker can be found in the `acceptability.AcceptabilityChecker` section below.
 - `analysis.py`: Abstract base classes for compiling the results of crowdsourcing runs.
-- `frontend.py`: Method for compiling the frontend code of crowdsourcing tasks.
+- `frontend.py`: Method for compiling the frontend code of crowdsourcing tasks. See `analysis.py` section below.
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/utils/README.md:6]: Hmm what's the link between `frontend.py` and `analysis.py` here?

File: parlai/crowdsourcing/utils/README.md [hunk 2]
```diff
@@ -14,3 +14,54 @@
...
+
+Contains abstract classes that provide the basic functionalities for compiling data from a Mephsito task.
+Mephisto provides two interfaces for retrieving its crowdsourced data; `MephistoDB` and `DataBrowser`.
+Using `AbstractResultsCompiler` you do not need to directly interact with these two modules---it provides an abstraction on top of these two.
+This class has methods such as `get_task_data` and `get_task_units` which handles interacting with Mephisto abstractions.
+For compiling your dataset from your crowdsourced Mephisto task, you need to extend this class and implement the following methods:
+
+* `compile_results` that returns a python *dictionary* (key-value pairs) or a pandas *dataframe*. We assume that, each unit of the crowdsourcing task (for example, annotation or conversation) has a unique id. This id is the key in the former case (the data from the unit is the value) and the row id for the latter.
+
+* (optoinal) `is_unit_acceptable` helps with simple filtering and data clean up. It receives the data from a unit of work and returns a boolean. We discrad this unit if it returns `False`.
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/utils/README.md:28]: nit: typos in "optional", "discard"

File: parlai/crowdsourcing/utils/README.md [hunk 3]
```diff
@@ -14,3 +14,54 @@
...
+
+## `analysis.py`
+
+Contains abstract classes that provide the basic functionalities for compiling data from a Mephsito task.
+Mephisto provides two interfaces for retrieving its crowdsourced data; `MephistoDB` and `DataBrowser`.
+Using `AbstractResultsCompiler` you do not need to directly interact with these two modules---it provides an abstraction on top of these two.
+This class has methods such as `get_task_data` and `get_task_units` which handles interacting with Mephisto abstractions.
+For compiling your dataset from your crowdsourced Mephisto task, you need to extend this class and implement the following methods:
+
+* `compile_results` that returns a python *dictionary* (key-value pairs) or a pandas *dataframe*. We assume that, each unit of the crowdsourcing task (for example, annotation or conversation) has a unique id. This id is the key in the former case (the data from the unit is the value) and the row id for the latter.
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/utils/README.md:26]: In the cases where `.compile_results()` returns a pandas DataFrame, each row actually indexes one turn of one conversation, not one conversation itself

File: parlai/crowdsourcing/utils/README.md [hunk 4]
```diff
@@ -14,3 +14,54 @@
...
+from parlai.crowdsourcing.utils.analysis import AbstractResultsCompiler
+
+class MyResultsCompiler(AbstractResultsCompiler):
+    def is_unit_acceptable(self, unit_data):
+        return unit_data['status'] ==  'accepted'
+
+    def compile_results(self):
+        data = dict()
+        for work_unit in self.get_task_data():
+            unit_id = work_unit.pop('ID)
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/utils/README.md:54]: hmm do we want to mutate `work_unit` here by popping the ID?

File: parlai/crowdsourcing/utils/README.md [hunk 5]
```diff
@@ -14,3 +14,54 @@
 - Add the code for this in `.check_messages()`, inside a `if 'check_name' in violation_types:` condition
 - Add the name of the check to `self.ALL_VIOLATION_TYPES`; otherwise, this check will not be recognized, and an error will be raised if the check is specified when calling `.check_messages()`!
 - To use the check: add the name of the check to the `violation_types` arg when calling `.check_messages()`
+
+## `analysis.py`
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/utils/README.md:18]: Great to have this section! Long overdue :)

File: parlai/crowdsourcing/utils/analysis.py [hunk 6]
```diff
@@ -46,11 +47,30 @@ def setup_args(cls):
             default='csv',
             help='Output format for results data',
         )
+        parser.add_argument(
+            '--task-name', type=str, help='Name of the Mephisto task to open'
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/utils/analysis.py:51]: 👍

File: parlai/crowdsourcing/utils/analysis.py [hunk 7]
```diff
@@ -62,6 +82,44 @@ def get_results_path_base(self) -> str:
             f'{self.__class__.__name__}__{now.strftime("%Y%m%d_%H%M%S")}',
         )
 
+    def get_worker_name(self, worker_id: str) -> str:
+        """
+        Gets the global (AWS) id of a worker from their Mephisto worker_id.
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/utils/analysis.py:87]: Hmm, the ID won't relate to AWS here if we're not using MTurk for our Mephisto job

File: parlai/crowdsourcing/utils/analysis.py [hunk 8]
```diff
@@ -62,6 +82,44 @@ def get_results_path_base(self) -> str:
...
+    def get_data_from_unit(self, unit: Unit) -> Dict[str, Any]:
+        """
+        Retrieves task data for a single unit.
+        """
+        try:
+            data_browser = self.get_mephisto_data_browser()
+            return data_browser.get_data_from_unit(unit)
+        except (IndexError, AssertionError):
+            logging.warning(
+                f'Skipping unit {unit.db_id}. No message found for this unit.'
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/utils/analysis.py:108]: Would be useful to print the error here (or else add a TODO to do so later)

File: tests/crowdsourcing/test_analysis.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+try:
+    from mephisto.data_model.unit import Unit
+
+    class MockUnit(Unit):
+        pass
+
+
+except ModuleNotFoundError:
+
+    class MockUnit:
```
Comment (EricMichaelSmith) [inline tests/crowdsourcing/test_analysis.py:21]: Hmm if we can still define a `MockUnit` successfully here even if we can't import `mephisto`, what's the advantage to importing `mephisto` in the first place?

File: tests/crowdsourcing/test_analysis.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        return kwargs.get("row")
+
+    def cache_result(self, cls, loaded_val):
+        pass
+
+    def get_worker_name(self, worker_id: str):
+        return {'worker_name': f'UNITTEST_MOCK_{worker_id}'}
+
+
+class MockMephistoBrowser:
```
Comment (EricMichaelSmith) [inline tests/crowdsourcing/test_analysis.py:48]: Really great to have this - this'll make testing of future crowdsourcing analysis scripts much easier :)

**Review by EricMichaelSmith — COMMENTED** [review]
Wow, really great to have this - thanks for pushing the PR out on this! The `MockResultsCompiler` in particular will be super valuable. I have a few nits + questions

**Review by JackUrb — APPROVED** [review]

**Comment by JackUrb** [comment]
The integrations with the Mephisto side make sense to me here, though Eric has a number of good questions and points still

File: parlai/crowdsourcing/utils/README.md (hunk 4 above)
Comment (mojtaba-komeili) [reply parlai/crowdsourcing/utils/README.md:54]: Yeah, but we are not reusing `work_unit` objects after this. We could keep the id but then it is our key for retrieving some content that includes the key itself. It's a bit of redundancy.

File: parlai/crowdsourcing/utils/analysis.py [hunk 11]
```diff
@@ -76,33 +134,57 @@ def is_unit_acceptable(self, unit_data: Dict[str, Any]) -> bool:
         return True
 
     @abstractmethod
-    def compile_results(self) -> pd.DataFrame:
+    def compile_results(self) -> Union[pd.DataFrame, Dict[str, Any]]:
```
[the remark this answers is not in the document]
Comment (mojtaba-komeili) [reply parlai/crowdsourcing/utils/analysis.py:137]: Completely agree.

File: parlai/crowdsourcing/utils/analysis.py (hunk 8 above)
Comment (mojtaba-komeili) [reply parlai/crowdsourcing/utils/analysis.py:108]: I added print of the error message.

File: tests/crowdsourcing/test_analysis.py (hunk 9 above)
Comment (mojtaba-komeili) [reply tests/crowdsourcing/test_analysis.py:21]: It was just for data type check later. Although I agree that make the code look messy.

File: parlai/crowdsourcing/utils/README.md [hunk 12]
```diff
@@ -23,9 +23,9 @@ Using `AbstractResultsCompiler` you do not need to directly interact with these
 This class has methods such as `get_task_data` and `get_task_units` which handles interacting with Mephisto abstractions.
 For compiling your dataset from your crowdsourced Mephisto task, you need to extend this class and implement the following methods:
 
-* `compile_results` that returns a python *dictionary* (key-value pairs) or a pandas *dataframe*. We assume that, each unit of the crowdsourcing task (for example, annotation or conversation) has a unique id. This id is the key in the former case (the data from the unit is the value) and the row id for the latter.
+* `compile_results` that returns a python *dictionary* (key-value pairs) or a pandas *dataframe*. We assume that, each unit of the crowdsourcing task (for example, annotation or conversation) has a unique id. In the json formnat, this id is the key for the entry that keeps dialogue data for that conversation. If the format is a dataframe, the convension is to have each row of the dataframe keep the data for a single utterance (interaction). Hence, the conversation id needs to be stored in a column for distinguishing the data from different dialogues.
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/utils/README.md:26]: nit: typo in "format"

File: parlai/crowdsourcing/utils/README.md [hunk 13]
```diff
@@ -23,9 +23,9 @@ Using `AbstractResultsCompiler` you do not need to directly interact with these
...
 
-* (optoinal) `is_unit_acceptable` helps with simple filtering and data clean up. It receives the data from a unit of work and returns a boolean. We discrad this unit if it returns `False`.
+* (optional) `is_unit_acceptable` helps with simple filtering and data clean up. It receives the data from a unit of work and returns a boolean. We discrad this unit if it returns `False`.
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/utils/README.md:28]: nit: typo in "discard"

File: parlai/crowdsourcing/utils/README.md (hunk 4 above)
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/utils/README.md:54]: Okay, fair, I don't have a strong preference here. But can you add the closing `'` after `ID`?

File: parlai/crowdsourcing/utils/analysis.py [hunk 14]
```diff
@@ -136,9 +140,15 @@ def is_unit_acceptable(self, unit_data: Dict[str, Any]) -> bool:
...
+        Method for returning the final results as a datafram or a json.
+
+        For Dict output each key is a unique identifier (eg Assignment ID) for a unit of
+        crowdsourcing work. The data for that unit is stored in the value as dictionary.
+
+        Each row of the dataframe consists of one utterance of one conversation, or crowdsourcing interaction.
 
-        Each row of the dataframe consists of one utterance of one conversation.
+        NOTE: Preference for new projects is Dict output (see the TODO below).
+        TODO: Only support Dict. Deprecate ` pd.DataFrame` when no other code is relyig on it.
```
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/utils/analysis.py:151]: Yeah, this is great! But maybe fix the typo in "relying"?

File: tests/crowdsourcing/test_analysis.py (hunk 9 above)
Comment (EricMichaelSmith) [reply tests/crowdsourcing/test_analysis.py:21]: Okay cool, makes sense :) Maybe you could add a comment saying that?

**Pull request merged**